### ISSUE TITLE
Crash during plan execution for local endpoint

### ISSUE THREAD

**Issue by jpeeler** [hashicorp/terraform-provider-aws#18187]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version
Terraform v0.14.8
+ provider registry.terraform.io/hashicorp/aws v3.32.0

### Affected Resource(s)

* aws_vpc

### Terraform Configuration Files

Please include all Terraform configurations required to reproduce the bug. Bug reports without a functional reproduction may be closed without investigation.

```hcl
locals {
  localstack_endpoint = "http://localhost:4566"
}

provider "aws" {
  region = "us-east-1"
  endpoints {
    accessanalyzer = local.localstack_endpoint
    acm = local.localstack_endpoint
    acmpca = local.localstack_endpoint
    amplify = local.localstack_endpoint
    apigateway = local.localstack_endpoint
    applicationautoscaling = local.localstack_endpoint
    applicationinsights = local.localstack_endpoint
    appmesh = local.localstack_endpoint
    appstream = local.localstack_endpoint
    appsync = local.localstack_endpoint
    athena = local.localstack_endpoint
    autoscaling = local.localstack_endpoint
    autoscalingplans = local.localstack_endpoint
    backup = local.localstack_endpoint
    batch = local.localstack_endpoint
    budgets = local.localstack_endpoint
    cloud9 = local.localstack_endpoint
    cloudformation = local.localstack_endpoint
    cloudfront = local.localstack_endpoint
    cloudhsm = local.localstack_endpoint
    cloudsearch = local.localstack_endpoint
    cloudtrail = local.localstack_endpoint
    cloudwatch = local.localstack_endpoint
    cloudwatchevents = local.localstack_endpoint
    cloudwatchlogs = local.localstack_endpoint
    codeartifact = local.localstack_endpoint
    codebuild = local.localstack_endpoint
    codecommit = local.localstack_endpoint
    codedeploy = local.localstack_endpoint
    codepipeline = local.localstack_endpoint
    codestarconnections = local.localstack_endpoint
    #codestarnotifications = local.localstack_endpoint
    cognitoidentity = local.localstack_endpoint
    cognitoidp = local.localstack_endpoint
    configservice = local.localstack_endpoint
    connect = local.localstack_endpoint
    cur = local.localstack_endpoint
    dataexchange = local.localstack_endpoint
    datapipeline = local.localstack_endpoint
    datasync = local.localstack_endpoint
    dax = local.localstack_endpoint
    devicefarm = local.localstack_endpoint
    directconnect = local.localstack_endpoint
    dlm = local.localstack_endpoint
    dms = local.localstack_endpoint
    docdb = local.localstack_endpoint
    ds = local.localstack_endpoint
    dynamodb = local.localstack_endpoint
    ec2 = local.localstack_endpoint
    ecr = local.localstack_endpoint
    ecrpublic = local.localstack_endpoint
    ecs = local.localstack_endpoint
    efs = local.localstack_endpoint
    eks = local.localstack_endpoint
    elasticache = local.localstack_endpoint
    elasticbeanstalk = local.localstack_endpoint
    elastictranscoder = local.localstack_endpoint
    elb = local.localstack_endpoint
    emr = local.localstack_endpoint
    emrcontainers = local.localstack_endpoint
    es = local.localstack_endpoint
    firehose = local.localstack_endpoint
    fms = local.localstack_endpoint
    forecast = local.localstack_endpoint
    fsx = local.localstack_endpoint
    gamelift = local.localstack_endpoint
    glacier = local.localstack_endpoint
    globalaccelerator = local.localstack_endpoint
    glue = local.localstack_endpoint
    guardduty = local.localstack_endpoint
    greengrass = local.localstack_endpoint
    iam = local.localstack_endpoint
    identitystore = local.localstack_endpoint
    imagebuilder = local.localstack_endpoint
    inspector = local.localstack_endpoint
    iot = local.localstack_endpoint
    iotanalytics = local.localstack_endpoint
    iotevents = local.localstack_endpoint
    kafka = local.localstack_endpoint
    kinesis = local.localstack_endpoint
    kinesisanalytics = local.localstack_endpoint
    kinesisanalyticsv2 = local.localstack_endpoint
    kinesisvideo = local.localstack_endpoint
    kms = local.localstack_endpoint
    lakeformation = local.localstack_endpoint
    lambda = local.localstack_endpoint
    lexmodels = local.localstack_endpoint
    licensemanager = local.localstack_endpoint
    lightsail = local.localstack_endpoint
    macie = local.localstack_endpoint
    macie2 = local.localstack_endpoint
    managedblockchain = local.localstack_endpoint
    marketplacecatalog = local.localstack_endpoint
    mediaconnect = local.localstack_endpoint
    mediaconvert = local.localstack_endpoint
    medialive = local.localstack_endpoint
    mediapackage = local.localstack_endpoint
    mediastore = local.localstack_endpoint
    mediastoredata = local.localstack_endpoint
    mq = local.localstack_endpoint
    mwaa = local.localstack_endpoint
    neptune = local.localstack_endpoint
    networkfirewall = local.localstack_endpoint
    networkmanager = local.localstack_endpoint
    opsworks = local.localstack_endpoint
    organizations = local.localstack_endpoint
    outposts = local.localstack_endpoint
    personalize = local.localstack_endpoint
    pinpoint = local.localstack_endpoint
    pricing = local.localstack_endpoint
    #prometheusservice = local.localstack_endpoint
    qldb = local.localstack_endpoint
    quicksight = local.localstack_endpoint
    ram = local.localstack_endpoint
    rds = local.localstack_endpoint
    redshift = local.localstack_endpoint
    resourcegroups = local.localstack_endpoint
    resourcegroupstaggingapi = local.localstack_endpoint
    route53 = local.localstack_endpoint
    route53domains = local.localstack_endpoint
    route53resolver = local.localstack_endpoint
    s3 = local.localstack_endpoint
    s3control = local.localstack_endpoint
    s3outposts = local.localstack_endpoint
    sagemaker = local.localstack_endpoint
    sdb = local.localstack_endpoint
    secretsmanager = local.localstack_endpoint
    securityhub = local.localstack_endpoint
    serverlessrepo = local.localstack_endpoint
    servicecatalog = local.localstack_endpoint
    servicediscovery = local.localstack_endpoint
    servicequotas = local.localstack_endpoint
    ses = local.localstack_endpoint
    shield = local.localstack_endpoint
    signer = local.localstack_endpoint
    sns = local.localstack_endpoint
    sqs = local.localstack_endpoint
    ssm = local.localstack_endpoint
    ssoadmin = local.localstack_endpoint
    stepfunctions = local.localstack_endpoint
    storagegateway = local.localstack_endpoint
    sts = local.localstack_endpoint
    swf = local.localstack_endpoint
    synthetics = local.localstack_endpoint
    timestreamwrite = local.localstack_endpoint
    transfer = local.localstack_endpoint
    waf = local.localstack_endpoint
    wafregional = local.localstack_endpoint
    wafv2 = local.localstack_endpoint
    worklink = local.localstack_endpoint
    workmail = local.localstack_endpoint
    workspaces = local.localstack_endpoint
    xray = local.localstack_endpoint
  }

  skip_credentials_validation = true
  skip_metadata_api_check     = true
  skip_requesting_account_id  = true
  s3_force_path_style         = true
}

module "vpc" {
  source = "terraform-aws-modules/vpc/aws"

  name = "my-vpc"
  cidr = "10.0.0.0/16"

  azs             = ["eu-west-1a", "eu-west-1b", "eu-west-1c"]
  private_subnets = ["10.0.1.0/24", "10.0.2.0/24", "10.0.3.0/24"]
  public_subnets  = ["10.0.101.0/24", "10.0.102.0/24", "10.0.103.0/24"]

  # this along with other endpoint enablements causes a crash
  enable_s3_endpoint = true
}
```

### Debug Output
(using my own binary on an m1, but I've duplicated on Linux too)
$ terraform-arm plan

Error: rpc error: code = Canceled desc = context canceled

panic: runtime error: index out of range [0] with length 0
2021-03-18T11:22:33.427-0400 [DEBUG] plugin.terraform-provider-aws: 
2021-03-18T11:22:33.427-0400 [DEBUG] plugin.terraform-provider-aws: goroutine 79 [running]:
2021-03-18T11:22:33.428-0400 [DEBUG] plugin.terraform-provider-aws: github.com/terraform-providers/terraform-provider-aws/aws.dataSourceAwsVpcEndpointServiceRead(0x14001e28480, 0x108265c00, 0x1400017d700, 0x108f68e28, 0x0)
2021-03-18T11:22:33.428-0400 [DEBUG] plugin.terraform-provider-aws: 	/Users/jpeeler/src/terraform-provider-aws/aws/data_source_aws_vpc_endpoint_service.go:192 +0x1230
2021-03-18T11:22:33.428-0400 [DEBUG] plugin.terraform-provider-aws: github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0x14000414840, 0x1093048c8, 0x140016e1480, 0x14001e28480, 0x108265c00, 0x1400017d700, 0x0, 0x0, 0x0)
2021-03-18T11:22:33.428-0400 [DEBUG] plugin.terraform-provider-aws: 	/Users/jpeeler/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.4.4/helper/schema/resource.go:290 +0x5c
2021-03-18T11:22:33.428-0400 [DEBUG] plugin.terraform-provider-aws: github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).ReadDataApply(0x14000414840, 0x1093048c8, 0x140016e1480, 0x14000f42620, 0x108265c00, 0x1400017d700, 0x1400017d700, 0x14000f42620, 0x0, 0x0)
2021-03-18T11:22:33.428-0400 [DEBUG] plugin.terraform-provider-aws: 	/Users/jpeeler/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.4.4/helper/schema/resource.go:498 +0xcc
2021-03-18T11:22:33.428-0400 [DEBUG] plugin.terraform-provider-aws: github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadDataSource(0x140006e24e0, 0x1093048c8, 0x140016e1480, 0x14000e6d080, 0x140016e1480, 0x10318f0ac, 0x108ad4d40)
2021-03-18T11:22:33.428-0400 [DEBUG] plugin.terraform-provider-aws: 	/Users/jpeeler/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.4.4/helper/schema/grpc_provider.go:1105 +0x3dc
2021-03-18T11:22:33.428-0400 [DEBUG] plugin.terraform-provider-aws: github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).ReadDataSource(0x1400172acc0, 0x109304970, 0x140016e1480, 0x140000b3bd0, 0x1400172acc0, 0x1030fe220, 0x108b42860)
2021-03-18T11:22:33.428-0400 [DEBUG] plugin.terraform-provider-aws: 	/Users/jpeeler/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.2.1/tfprotov5/server/server.go:247 +0xc4
2021-03-18T11:22:33.428-0400 [DEBUG] plugin.terraform-provider-aws: github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadDataSource_Handler(0x108d742c0, 0x1400172acc0, 0x109304970, 0x14001422b70, 0x14000716fc0, 0x0, 0x109304970, 0x14001422b70, 0x1400183b2c0, 0x124)
2021-03-18T11:22:33.428-0400 [DEBUG] plugin.terraform-provider-aws: 	/Users/jpeeler/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.2.1/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:416 +0x1c8
2021-03-18T11:22:33.429-0400 [DEBUG] plugin.terraform-provider-aws: google.golang.org/grpc.(*Server).processUnaryRPC(0x14000050fc0, 0x1093242f8, 0x14000bd0600, 0x14001e13c00, 0x140010569c0, 0x10c65ce30, 0x0, 0x0, 0x0)
2021-03-18T11:22:33.429-0400 [DEBUG] plugin.terraform-provider-aws: 	/Users/jpeeler/go/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:1194 +0x3e8
2021-03-18T11:22:33.429-0400 [DEBUG] plugin.terraform-provider-aws: google.golang.org/grpc.(*Server).handleStream(0x14000050fc0, 0x1093242f8, 0x14000bd0600, 0x14001e13c00, 0x0)
2021-03-18T11:22:33.429-0400 [DEBUG] plugin.terraform-provider-aws: 	/Users/jpeeler/go/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:1517 +0xa50
2021-03-18T11:22:33.429-0400 [DEBUG] plugin.terraform-provider-aws: google.golang.org/grpc.(*Server).serveStreams.func1.2(0x14000bb8200, 0x14000050fc0, 0x1093242f8, 0x14000bd0600, 0x14001e13c00)
2021-03-18T11:22:33.429-0400 [DEBUG] plugin.terraform-provider-aws: 	/Users/jpeeler/go/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:859 +0x94
2021-03-18T11:22:33.429-0400 [DEBUG] plugin.terraform-provider-aws: created by google.golang.org/grpc.(*Server).serveStreams.func1
2021-03-18T11:22:33.429-0400 [DEBUG] plugin.terraform-provider-aws: 	/Users/jpeeler/go/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:857 +0x1f8
2021-03-18T11:22:33.430-0400 [DEBUG] plugin: plugin process exited: path=/Users/jpeeler/src/terraform-provider-aws/terraform-provider-aws pid=27910 error="exit status 2"
2021/03/18 11:22:33 [TRACE] vertex "module.vpc.data.aws_vpc_endpoint_service.s3[0]": visit complete
2021/03/18 11:22:33 [TRACE] vertex "module.vpc.data.aws_vpc_endpoint_service.s3": dynamic subgraph encountered errors
2021/03/18 11:22:33 [TRACE] vertex "module.vpc.data.aws_vpc_endpoint_service.s3": visit complete
2021/03/18 11:22:33 [TRACE] vertex "module.vpc.data.aws_vpc_endpoint_service.s3 (expand)": dynamic subgraph encountered errors
2021/03/18 11:22:33 [TRACE] vertex "module.vpc.data.aws_vpc_endpoint_service.s3 (expand)": visit complete
2021/03/18 11:22:33 [TRACE] dag/walk: upstream of "module.vpc.aws_vpc_endpoint.s3 (expand)" errored, so skipping
2021/03/18 11:22:33 [TRACE] dag/walk: upstream of "module.vpc.output.vpc_endpoint_s3_id (expand)" errored, so skipping
2021/03/18 11:22:33 [TRACE] dag/walk: upstream of "module.vpc.aws_vpc_endpoint_route_table_association.intra_s3 (expand)" errored, so skipping
2021/03/18 11:22:33 [TRACE] dag/walk: upstream of "module.vpc.output.vpc_endpoint_s3_pl_id (expand)" errored, so skipping
2021/03/18 11:22:33 [TRACE] dag/walk: upstream of "module.vpc.aws_vpc_endpoint_route_table_association.public_s3 (expand)" errored, so skipping
2021/03/18 11:22:33 [TRACE] dag/walk: upstream of "module.vpc.aws_vpc_endpoint_route_table_association.private_s3 (expand)" errored, so skipping
2021-03-18T11:22:33.430-0400 [WARN]  plugin.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2021/03/18 11:22:33 [TRACE] dag/walk: upstream of "provider[\"registry.terraform.io/hashicorp/aws\"] (close)" errored, so skipping
2021/03/18 11:22:33 [TRACE] dag/walk: upstream of "module.vpc (close)" errored, so skipping
2021/03/18 11:22:33 [TRACE] dag/walk: upstream of "meta.count-boundary (EachMode fixup)" errored, so skipping
2021/03/18 11:22:33 [TRACE] dag/walk: upstream of "root" errored, so skipping
2021/03/18 11:22:33 [INFO] backend/local: plan operation completed
2021/03/18 11:22:33 [TRACE] statemgr.Filesystem: removing lock metadata file .terraform.tfstate.lock.info
2021/03/18 11:22:33 [TRACE] statemgr.Filesystem: unlocking terraform.tfstate using fcntl flock
2021-03-18T11:22:33.431-0400 [DEBUG] plugin: plugin exited

### Panic Output

https://gist.github.com/jpeeler/45d9d7faad87f0ec7409ddc82545d073

### Expected Behavior

Terraform plan should execute successfully.

### Actual Behavior

Terraform crashed.

### Steps to Reproduce

1. `terraform plan`

### Important Factoids

I'm running [localstack](https://github.com/localstack/localstack) and have changed all endpoints to utilize it as shown in main.tf.

### References

This issue is slightly related as it crashes in the same area, but the associated PR didn't fix the problem:

https://github.com/hashicorp/terraform-provider-aws/issues/17419

**Comment by whummer** [comment]
Thanks for reporting @jpeeler - please note that this was likely related to some incorrect responses returned from LocalStack. This has now been fixed in  localstack/localstack#3764 - can you please pull the `localstack/localstack:latest` image and give it another try? Thanks for your help.

**Comment by jpeeler** [comment]
@whummer I've confirmed the crash is fixed - thanks for the quick fix!

Normally I would leave this open as a plugin should never crash, regardless of the response from AWS. However, I see there are 2700 open issues in this repo. I highly doubt that a crash that only occurs with localstack is going to be invested in anytime soon if ever.
